### PR TITLE
Factory to populate Config object from well-known locations

### PR DESCRIPTION
**Pull request by jerrinot**
It allows to load config from an external file and then customize
it programmatically. This is useful when embedding Hazelcast into
managed environments: When a runtime container wants to create an
instance as if it was created via `Hazelcast.newHazelcastInstance()`,
but it wants to inject an instance of `ManagedContext` into Hazelcast
configuration.

Without this capability all containers have to duplicate the `Config`
location strategy: "First try this, then that, etc."

Related to https://github.com/spring-projects/spring-boot/issues/19487

TODO: Tests

### CODE REVIEW

**Comment by jerrinot** [comment]
I realize it's VERY late in the dev cycle, so this is more for a consideration. I wonder what does Zoltan think about this change.

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 1]
```diff
@@ -2649,4 +2653,42 @@ public String toString() {
...
     }
+
+    /**
+     * Populate Hazelcast <code>Config</code> object from an external configuration file.
+     * It tries to load Hazelcast configuration from a list of well-known locations.
+     * When no location contains Hazelcast configuration then it returns default.
+     *
+     * @return Config created from a file when exists, otherwise default.
+     */
+    public static Config load() {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/config/Config.java]: not a great name

**Comment by jerrinot** [comment]
run-lab-run

**Comment by jerrinot** [comment]
Additional reasoning behind this change (copy and pasted from a conversation with @mmedenjak)

A runtime container typically wants to either:
1. start an instance by using programmatic configuration. `new Config().setFoo().setBar().setManagedContext()`
2. start new instance with a specific configuration file.
3. start a new instance as if it was started by `Hazelcast.newHazelcastInstance()`, but amend the `Config` with an instance of `ManagedContext`

#1 works without any problem, obviously.
#2 works OK-ish. You have to create either a `YamlConfigBuilder` or `XmlConfigBuilder`, Construct the `Config` object and then set the `ManagedContext`. Ideally I think we should hide the builder objects hidden from users. Just pass a path to a factory method on the Config object. The Config object can detect what file is that and create the Xml/Yaml Builder behind the scene. But perhaps there is a reason to have the builders exposed. Perhaps constructing from `InputStream` with a config?
#3 is currently complicated. You need to copy&paste the config file location strategy from Hazelcast to the container, create the instance of `Config` and then inject the `ManagedContext`. + if we change the strategy then the container code has to be changed as well. for a consistent experience.

**Comment by mmedenjak** [comment]
Does this fix https://github.com/hazelcast/hazelcast/issues/16400?

**Comment by jerrinot** [comment]
@mmedenjak it does not.

**Comment by jerrinot** [comment]
run-lab-run

**Comment by blazember** [comment]
The builders are public for the reason you mentioned: it's a flexible API letting the users to build the config from any source, mostly through the `InputStream` constructor. 

Exposing the locator logic on a public API is a good step I believe. Since we have the convenient `Config` classes (`FileSystemXmlConfig` & co), I'd move it to something like `DeclarativeConfig` instead of adding it to `Config` directly. It could be used this way: `new DeclarativeConfig().setManagedContext();`.

As an additional thought, since we have convenient `Config` classes, it might make sense to create one for `InputStream`s too, so that we can make the builders private. 

WDYT?

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 2]
```diff
@@ -2649,4 +2653,42 @@ public String toString() {
...
+        validateSuffixInSystemProperty(SYSPROP_MEMBER_CONFIG);
+
+        XmlConfigLocator xmlConfigLocator = new XmlConfigLocator();
+        YamlConfigLocator yamlConfigLocator = new YamlConfigLocator();
+
+        Config config;
+        if (xmlConfigLocator.locateFromSystemProperty()) {
+            // 1. Try loading XML config from the configuration provided in system property
+            config = new XmlConfigBuilder(xmlConfigLocator).build();
+
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/config/Config.java]: Are these empty lines required?

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 3]
```diff
@@ -2649,4 +2653,42 @@ public String toString() {
                 + ", metricsConfig=" + metricsConfig
                 + '}';
     }
+
+    /**
+     * Populate Hazelcast <code>Config</code> object from an external configuration file.
+     * It tries to load Hazelcast configuration from a list of well-known locations.
+     * When no location contains Hazelcast configuration then it returns default.
+     *
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/config/Config.java]: Can you add a note that the same mechanism is used when you call Hazelcast.newHazelcastInstance() ?

File: hazelcast/src/main/java/com/hazelcast/config/Config.java (hunk 1 above)
Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java]: I have no problem with the name. It explains that the config is loaded and this implies some files being checked. Unless you can come up with a better name, I would leave it as it is.

**Comment by pveentjer** [comment]
I like this API. It is simple from a user perspective. It doesn't expose anything that shouldn't be exposed. 

So I give my thumbs up.

image

I would like to see a note that calling
```
HazelcastInstance hz = Hazelcast.newHazelcastInstance()
```

Is exactly the same as:
```
Config config = Config.load();
HazelcastInstance hz = Hazelcast.newHazelcastInstance(config);
```

**Comment by Holmistr** [comment]
run-lab-run

**Comment by leszko** [comment]
Closing this one. I'll open a new PR with this in a moment.

**Pull request closed without merging**